### PR TITLE
Introduce the `dataChangeEventDisabled` property

### LINKED ISSUE #141: The component works slow on large documents
Similar issues:
https://github.com/ckeditor/ckeditor5-vue/issues/42
And:
https://github.com/ckeditor/ckeditor4/issues/449

But then for CKEditor 5 and angular. I am experiencing the same issues and wonder how to fix this? I guess this has been brought up before but I can't seem to find any related topics on this.

This can easily be re-produced by opening the default template for angular ckeditor:
https://stackblitz.com/edit/angular-ckeditor
And paste large amount of text in it. Go to https://www.lipsum.com/, put paragraphs on 150 and just start copy-pasting that text into the edtior. Around 5 or 6 times it becomes slower. The more lines you have, the more slower it becomes.

I have tried to remove the `[data]` property so that there is no two-way binding, but I guess, just like in the vue project, there is an issue with the on-change detection.

### PR DESCRIPTION
**Pull request by Ruud-cb**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduced the `dataChangeEventDisabled` property that will enable performance boost for large documents. Closes #141.

---

### Additional information

*For example – encountered issues, assumptions you had to make, other affected tickets, etc.*

### CODE REVIEW

**Comment by oleq** [comment]
Hi, @Ruud-cb! Can you tell us more about what problem does this PR actually solve? Some use–case or maybe a demo?

**Comment by Ruud-cb** [comment]
Hi @oleq, follow the reproduction steps in the issue, use the demo-form component. You will notice extreme lag after inserting a lot of text. Then change `public disabledOnChangeEvent: boolean = false;` to `true`. The lag during typing/pasting text is solved. Still lag noticeable during leaving the textbox unfortunately, but at some point the new data has to be sent to `ngModel`.

**Comment by ma2ciek** [comment]
Thanks @Ruud-cb for the PR.

> The lag during typing/pasting text is solved. Still lag noticeable during leaving the textbox unfortunately, but at some point the new data has to be sent to ngModel.

It's something we can do nothing about, the `editor.getData()` must be called at some point. However, I'd change the property name to something like `muteDataChangeEvent`, as it touches both, the `[(ngModel)]`'s control value accessor and the `<ckeditor (change)>`. Also, could you write a test for this change?

**Comment by ma2ciek** [comment]
> I could not find a way to get the text manually since Editor.getData() does not exist, so I execute cvaOnChange when the text editor blurs.

What do you mean? you can get the data with `editor.getData()`.

File: src/app/demo-form/demo-form.component.ts [hunk 1]
```diff
@@ -35,6 +37,9 @@ export class DemoFormComponent implements AfterViewInit {
 	}
 
 	public onSubmit() {
+		if ( this.disabledOnChangeEvent ) {
+			this.model.description = this.Editor.getData();
```
Comment (ma2ciek) [inline src/app/demo-form/demo-form.component.ts]: There's no such thing in the class constructor. You need to get the instance of this class instead.

**Comment by Ruud-cb** [comment]
> Thanks @Ruud-cb for the PR.
> 
> > The lag during typing/pasting text is solved. Still lag noticeable during leaving the textbox unfortunately, but at some point the new data has to be sent to ngModel.
> 
> It's something we can do nothing about, the `editor.getData()` must be called at some point. However, I'd change the property name to something like `muteDataChangeEvent`, as it touches both, the `[(ngModel)]`'s control value accessor and the `<ckeditor (change)>`. Also, could you write a test for this change?

I understand that `editor.getData()` must be called at some point, no problem for me that this happens on blur, it doesn't happen during typing which is great.


> > I could not find a way to get the text manually since Editor.getData() does not exist, so I execute cvaOnChange when the text editor blurs.
> 
> What do you mean? you can get the data with `editor.getData()`.

Sorry for the confusion, `Editor.getData()` was used in the demo component (or whichever component uses the ckeditor component). As far as I found that doesn't exist. As you commented on my code, indeed that does not exist on the class constructor. I forgot to delete it again because it's now not needed anymore in `onSubmit()` because `(blur)` will occur before the user clicks on `onSubmit()`, right?

It would be favorable if you can adjust the unit tests as it will take you significant less time.

**Comment by ma2ciek** [comment]
After a while I think about this, such change can be dangerous for big documents as they might be not saved for a long time, as now the saving is done only during the `blur` event. A debounce-based approach could be a better solution as this should work out of the box, without providing any unuseful properties. Although I'm not sure if that approach will work with Angular Reactive Forms too.

**Comment by Ruud-cb** [comment]
> After a while I think about this, such change can be dangerous for big documents as they might be not saved for a long time, as now the saving is done only during the `blur` event. A debounce-based approach could be a better solution as this should work out of the box, without providing any unuseful properties. Although I'm not sure if that approach will work with Angular Reactive Forms too.

True, although it is deliberately set by the developer, so one should know. Currently the problem is that no text can be typed in big documents smoothly. And to solve the 'losing data' problem, I would suggest developers to make use of `IComponentCanDeactivate`, which can give produce an warning when trying to reload, go back or exit the page.  Of course this is beyond the scope of this project.

**Comment by ma2ciek** [comment]
> True, although it is deliberately set by the developer, so one should know. Currently the problem is that no text can be typed in big documents smoothly. And to solve the 'losing data' problem, I would suggest developers to make use of IComponentCanDeactivate, which can produce a warning when trying to reload, go back or exit the page. Of course this is beyond the scope of this project

But I think that using the `debounce` effect is safer and it should allow the same editing experience. This is a better approach than "I know it can be an unsafe feature, but I have big documents so I have no options."

**Comment by Ruud-cb** [comment]
Alright, that's also possible. Only argument against that is that it might occur the user starts typing again exactly at the moment the debounce is fired off. Maybe both, whatever you want :).

**Comment by tgv1975** [comment]
Hi @ma2ciek, the build is failing on this due to unit test coverage being under 100%, so your changes have not been merged, the large document problem was not solved.

**Comment by ma2ciek** [comment]
I think that this change is no longer required as the `editor.getData()` only happens in the 2-way binding and it's a must there - see:

[https://github.com/ckeditor/ckeditor5-angular/blob/bdfabf453349004043c930e605e7ccd0954bc804/src/ckeditor/ckeditor.component.ts#L383-L393](https://github.com/ckeditor/ckeditor5-angular/blob/bdfabf453349004043c930e605e7ccd0954bc804/src/ckeditor/ckeditor.component.ts#L383-L393)

**Pull request closed without merging**